Improve GDL python support to work both ways

rmsds opened the pull request:
Only the possibility of using Python from GDL was being built, now the GDL python module is also being built, and linked into the correct location.

rmsds committed:
Fix python support for GDL

* python module is installed in an 'old' location which is not used by default

rmsds committed:
Fix Python Support

* also build the python_module (and matching description for the variant)
* bugfix the linking the python library from site-python to site-packages

Signed-off-by: Ricardo Silva <ricardo.silva@epfl.ch>

rmsds committed:
Merge branch 'releases/paien' into bugfix/paien/gdl_python

rmsds committed:
Merge branch 'releases/paien' into bugfix/paien/gdl_python

nrichart commented on var/spack/repos/builtin/packages/gdl/package.py:
```diff
@@ -127,3 +132,20 @@ def cmake_args(self):
+                'site-python')
+            dst = os.path.join(
+                self.spec.prefix.lib,
+                'python{0}'.format(self.spec['python'].version.up_to(2)),
```
That is the official spack way to find out where to install python stuff ?

rmsds replied:
I'll look into actually moving the file as it is probably safer in case something ever decides to look into site-python and it's not the correct python version.

> That is the official spack way to find out where to install python stuff ?

Probably not, I'll try to find out.

nrichart reviewed (commented):
In fact why do you bother to maintain the site-python and not in case it was installed in `site-python` juste move `site-python` as `site-packages/python-{version}` ?

rmsds committed:
Improvements

* use python defined property for the site_packages_dir
* move instead of symlinking the python module

Signed-off-by: Ricardo Silva <ricardo.silva@epfl.ch>

rmsds commented:
I've opted to move the files because it the `site_packages_dir` seems to be created even if it is left empty. Moving the files seems more robust.

rmsds commented:
Do not merge yet, I just noticed this package either builds the command line tool *or* the python module, but not both. I need some help with this one.

rmsds committed:
Patch GDL CMakeLists to build everything

* when building the Python module it would *only* build that
* others have solved it by creating a shared library and patching the CMakeLists:
  https://sources.debian.org/patches/gnudatalanguage/0.9.8-7/Create-a-shared-library.patch/
* this might get merged upstream:
  https://github.com/gnudatalanguage/gdl/issues/49

rmsds committed:
Bugfix patch and Python Module logic improvement

* syntax error on patch
* Only enable the Python module for the patched version

rmsds committed:
Fix build with and without python modules

* debian patch alone made it work for +python but broke it with ~python
* added a patch to always build as a shared library (otherwise got errors due to non relocatable stuff)
* renamed python variants to match the spack conventions

rmsds committed:
Improvements

* simplified python-embed + patch logic to reduce number of places where a version number is hardcoded
* bugfix after renaming of python-variants

rmsds committed:
Group related comments

rmsds committed:
Merge branch 'releases/paien' into bugfix/paien/gdl_python

alalazo commented on var/spack/repos/builtin/packages/gdl/package.py:
```diff
@@ -127,3 +150,21 @@ def cmake_args(self):
+                'site-python')
+            dst = os.path.join(
+                self.spec.prefix,
+                site_packages_dir)
```
In the `python` package I currently see:
```python
module.site_packages_dir = join_path(dependent_spec.prefix,
                                     self.site_packages_dir)
```
I wonder if we joined the `self.spec.prefix` 2 times here. Can you double check?

rmsds replied:
I don't really understand why (because I see the `site_packages_dir` being used as you say in other packages) but it works fine. I've reinstalled all 3 possible combinations of (+/~)(embed-)python without issues.

rmsds replied:
I think I know why it works, from `help(os.path.join)`:
> If any component is an absolute path, all previous path components will be discarded.

alalazo commented on var/spack/repos/builtin/packages/gdl/package.py:
```diff
@@ -46,7 +48,8 @@ class Gdl(CMakePackage):
     variant('openmp', default=True, description='Enable OpenMP')
     variant('proj', default=True, description='Enable LIBPROJ4')
-    variant('python', default=False, description='Enable Python')
+    variant('python-embed', default=False, description='Ability to embed Python within GDL')
```
Minor comment on naming: I'd prefer `embedded-python` or `embed-python` rather than `python-embed`. Not requested change, of course.

rmsds replied:
I think I know why it works, from `help(os.path.join)`:
> If any component is an absolute path, all previous path components will be discarded.

alalazo reviewed (commented):
LGTM. Can you please double check the comment below on `dst` and Python `site_packages_dir`?

rmsds committed:
Rename python-embed to embed-python variant

rmsds committed:
Bugfix: do not join the prefix twice

rmsds committed:
Python 3.3 testing is broken

* it seems six is no longer installs for Python 3.3
* https://github.com/spack/spack/pull/9098
